design/one-pager-merge-jets-proposal.md

```diff
@@ -0,0 +1,432 @@
+vpcIdSelector:
+  matchLabels:
+    app: ola
+```
+
+When we have Jet and classic providers, one does not know about the other's
+types. Hence, the only way to do references is to give the type of the
+referenced objct and the field path of the information on that type, something
+roughly like the following:
+
+```yaml
+vpcId: vpc-2314334
+vpcIdRef:
+  apiVersion: ec2.aws.crossplane.io/v1beta1
+  kind: VPC
+  fieldPath: metadata.annotations[crossplane.io/external-name]
+vpcIdSelector:
+  apiVersion: ec2.aws.crossplane.io/v1beta1
+  kind: VPC
+  fieldPath: metadata.annotations[crossplane.io/external-name]
+  matchLabels:
+    app: ola
+```
+
+While this referencing style is useful for many scenarios and being discussed
+[here][generic-references], it's not the same as how Kubernetes resources
+reference each other and requires more upfront information. We could possibly
+make it work like native with a default and let users optionally give different
+`kind`, `apiVersion` and `fieldPath`, there are other concerns with the approach
+listed in its discussion.
+
+In a single provider, users would keep using Kubernetes-native style referencing
+where they are only concerned with the name and/or labels of the referenced
+object, since there is a single type and it's defined in the same provider, the
+controller knows these details already.
+
+## Drawbacks
+
+### Code Difference
+
+Terrajet has a single controller implementation that you configure to work with
+different CRDs but the folder structure is the same as classic providers, i.e.
+`apis` and `pkg/controller` folders and the rest is exactly the same. The most
+apparent difference is that in classic providers, you have the controller
+`Setup` function [together with the implementation][classic-setup] of the
+controller whereas in Jet providers, there is only `Setup` [function][jet-setup]
+since its implementation lives in `terrajet` repository. In either case, the
+controller implementation satisfies the same
+[`ExternalClient`][externalclient-interface] interface from crossplane-runtime.
+
+The biggest difference in that regard is that the CRD structs in Jet providers
+need to satisfy an additional interface called
+[`Terraformed`][terraformed-interface] in order for the generic controller to
+work. So, even though the structure is same, you'll see that [the generic
+controller][generic-controller] makes extensive use of the functions defined in
+that interface where as classic implementations work with the type directly.
+
+### Maintenance Challenges
+
+This is probably the biggest difference between classic and Jet providers. If
+you run the debugger and put breakpoints in the code, you'll realize that Jet
+providers run the Terraform CLI in a temporary TF workspace folder instead of
+making HTTP calls to the cloud vendor. This makes it more challenging to debug
+Jet providers because you need to have a rough idea of how Terraform works, i.e.
```

Comment:
I do get the desire to increase adoption by increasing coverage, but this proposal is **incredibly disappointing and disheartening**. I've evangelized Crossplane at my company and we've been so interested in it and started implementing it because we've all experienced the suckage of automating Terraform and troubleshooting it and fixing it's state management.

I've worked at companies that would vito the continued use of Crossplane if it's going to shell out as that has negative security implications, not to mention it's an ugly smelly pattern, and like you said is a phenomenally awful debugging experience. Another company I've worked at would vito it _just_ because it's aiming for Terraform.

I would rather have to write my own where I need coverage than have to debug Terraform being used by Crossplane.

I can see combining the providers, but I can't believe we're considering moving towards Terraform and away from native. We will be looking for another solution if that's the case.

Reply:
>I've evangelized Crossplane at my company and we've been so interested in it and started implementing it because we've all experienced the suckage of automating Terraform and troubleshooting it and fixing it's state management.

I understand the frustration you've experienced with Terraform but it would be really helpful if you can point to specific details that we can relate to our usage of Terraform in Terrajet. I think quite a good portion of them may not be relevant since we're interacting with Terraform and its providers in a single and consistent way via its stable interfaces.

> I would rather have to write my own where I need coverage than have to debug Terraform being used by Crossplane.

While Terraform and Crossplane have architectural differences, some of the problems users experience with Terraform are because of the problem that both projects try to solve - building a declarative API on top of imperative calls of cloud vendors. So I think the classic providers are likely to have some users with experience similar to TF providers that are written manually once they get to that coverage level. To be fair, the set of problems are not limited to those common ones but without the specifics, it's hard to dismiss that the worries you've got will appear in classic providers as well.

> I can see combining the providers, but I can't believe we're considering moving towards Terraform and away from native.

We need to define what _native_ is in this context and I assume it's making direct SDK calls in the controller, i.e. what's defined as _classic_ in this doc. And as Crossplane maintainers, I think we it's safe to say that we're seeing manual implementation of the controllers as not sustainable due to the sheer number of APIs exposed by these cloud vendors. In fact, the cloud vendors themselves also come up with tooling to scale as mentioned in the doc like AWS Cloud Control, Azure Resource Manager. So, what we're betting on is that the next step after Terrajet will not be _classic_ but more likely these tools that, similar to Terrajet, sit in the middle of the controller and the final API calls. So we're moving away from classic is because vendors don't prefer it either and there isn't a really good way to scale their maintenance without a middle layer - be it Terrajet or vendor tools. At that point, it's hard to tell that making direct API calls is _native_ and others are not. More details [here](https://github.com/crossplane/crossplane/blob/master/design/design-doc-provider-strategy.md) about these tools in the strategy doc.

Reply:
@muvaf Thanks for the level headed response. I definitely spoke from a place of little knowledge of how Crossplane uses Terraform, that's on me.

> it would be really helpful if you can point to specific details that we can relate to our usage of Terraform in Terrajet. I think quite a good portion of them may not be relevant since we're interacting with Terraform and its providers in a single and consistent way via its stable interfaces.

It's been a while since I've been deep in the weeds with Terraform, but the kinds of problems I remember are:
* The troubleshooting and contribute-ablity issue: When I've automated Terraform it always involved shelling out to the binary. The problem is not just about running a debugger, it's also about following the logic path mentally. When there's a problem I'll often just go straight to reading the code, but the reasonableness of doing that shifts significantly with the issues surrounding having to jump to a new codebase, and I've repeatedly found myself feeling it's easier to stab in the dark until it works with Terraform than track logic through it's code and then track that back to the automation. It's not just about finding the right place in the Terraform code to pick up reading, it involves tracking down Terraform and Module versions and context switching between them as well. You can't "GoToDefinition". IMO, the path to debugging has crossed a critical threshold of maintainability.
* The lack of comparing the real-world to the intended-world before running operations, (instead, comparing to the state file.): - This may not be the case with Crossplane/Terrajet's usage of Terraform? - It sounds like you're probably familiar with the issues surrounding that state file getting out of sync due to an interrupted operations, etc.

> So I think the classic providers are likely to have some users with experience similar to TF providers that are written manually once they get to that coverage level.

I agree there'll be bugs in both forms of providers, but to me it's about the approachability of providing fixes. With one approach it's all right there in one place in the code, with the other it's a whole different paradigm.

> I think we it's safe to say that we're seeing manual implementation of the controllers as not sustainable

OK. I'm shifting my thinking...

> So, what we're betting on is that the next step after Terrajet will not be classic but more likely these tools that, similar to Terrajet, sit in the middle of the controller and the final API calls. So we're moving away from classic is because vendors don't prefer it either and there isn't a really good way to scale their maintenance without a middle layer - be it Terrajet or vendor tools. At that point, it's hard to tell that making direct API calls is native and others are not.

Oh, OK, I didn't followed this part. The consideration of using those middle layer APIs is kind of key though, right? - **That's hugely different** than what I understood this proposal to be, because the vendors will maintain those middle APIs (AWS Cloud Control, etc), which means we won't have to dubug them, we won't have to follow logic from Crossplane code through to Terrajet through to Terraform through to a Module's API calls. - We'll only have to follow logic inside Crossplane's code and imported libraries, and report issues with a vendor's API to that vendor.